Automattic/wp-calypso
# Checkout: Convert CheckoutThankYou to TypeScript

## Proposed Changes

This PR converts the `CheckoutThankYou` component and several of its dependencies to TypeScript with as few functional changes as possible.

This required one large refactor because the component had a very hard-to-read and highly dynamic method called `getComponentAndPrimaryPurchaseAndDomain()` that TypeScript wouldn't even allow to exist. The method should really be replaced but to simplify this PR, I retained the method's original functionality after removing its component class return value. The wrapper `PurchaseDetailsWrapper` now should handle the actual rendering it used to be used for.

<img width="963" alt="Screenshot 2023-05-10 at 4 03 50 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/8941db34-fa33-4c8f-ae43-6ccd543d586f">

## Testing Instructions

I added some tests but there are a lot of complicated logic patterns in this component that are hard to understand. I have a feeling that much of the code this component contains can be refactored or removed, but this step should simplify doing so.

For example, as far as I can tell we cannot ever render the [generic domain registration details component](https://github.com/Automattic/wp-calypso/blob/4903ab03c1ed33f0b759ae1312b99e3e5547f2e9/client/my-sites/checkout/checkout-thank-you/index.tsx#L947) because it is only rendered [if there is a receipt ID](https://github.com/Automattic/wp-calypso/blob/4903ab03c1ed33f0b759ae1312b99e3e5547f2e9/client/my-sites/checkout/checkout-thank-you/index.tsx#L733) and [a different component is rendered](https://github.com/Automattic/wp-calypso/blob/4903ab03c1ed33f0b759ae1312b99e3e5547f2e9/client/my-sites/checkout/checkout-thank-you/index.tsx#L589) when [a domain registration was purchased](https://github.com/Automattic/wp-calypso/blob/4903ab03c1ed33f0b759ae1312b99e3e5547f2e9/client/my-sites/checkout/checkout-thank-you/index.tsx#L484).

Some checkout URLs to test follow. Each one should look the same before and after this change:

- `/checkout/thank-you/no-site`
- `/thank-you/example.com/unknown-receipt` (where `example.com` should be changed to a real site hostname.)
- The thank-you page for a domain-only purchase (begin the purchase process at `/domains`).
- The thank-you page for a plan (begin the purchase process at `/plans`).
